core/src/SYCL/Kokkos_SYCL_GraphNodeKernel.hpp

```diff
@@ -0,0 +1,158 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#ifndef KOKKOS_SYCL_GRAPHNODEKERNEL_HPP
+#define KOKKOS_SYCL_GRAPHNODEKERNEL_HPP
+
+#include <Kokkos_Graph_fwd.hpp>
+
+#include <impl/Kokkos_GraphImpl.hpp>
+
+#include <Kokkos_Parallel.hpp>
+#include <Kokkos_Parallel_Reduce.hpp>
+#include <Kokkos_PointerOwnership.hpp>
+
+#include <SYCL/Kokkos_SYCL_GraphNode_Impl.hpp>
+
+namespace Kokkos {
+namespace Impl {
+
+template <typename PolicyType, typename Functor, typename PatternTag,
+          typename... Args>
+class GraphNodeKernelImpl<Kokkos::Experimental::SYCL, PolicyType, Functor,
+                          PatternTag, Args...>
+    : public PatternImplSpecializationFromTag<
+          PatternTag, Functor, PolicyType, Args...,
+          Kokkos::Experimental::SYCL>::type {
+ public:
+  using Policy       = PolicyType;
+  using graph_kernel = GraphNodeKernelImpl;
+  using base_t       = typename PatternImplSpecializationFromTag<
+      PatternTag, Functor, Policy, Args..., Kokkos::Experimental::SYCL>::type;
+
+  // TODO use the name and executionspace
+  template <typename PolicyDeduced, typename... ArgsDeduced>
+  GraphNodeKernelImpl(std::string, Kokkos::Experimental::SYCL const&,
+                      Functor arg_functor, PolicyDeduced&& arg_policy,
+                      ArgsDeduced&&... args)
+      : base_t(std::move(arg_functor), (PolicyDeduced &&) arg_policy,
+               (ArgsDeduced &&) args...) {}
+
+  template <typename PolicyDeduced>
+  GraphNodeKernelImpl(Kokkos::Experimental::SYCL const& exec_space,
+                      Functor arg_functor, PolicyDeduced&& arg_policy)
+      : GraphNodeKernelImpl("", exec_space, std::move(arg_functor),
+                            (PolicyDeduced &&) arg_policy) {}
+
+  ~GraphNodeKernelImpl() {
+    if (m_driver_storage) {
+      Kokkos::Experimental::SYCLDeviceUSMSpace().deallocate(m_driver_storage,
+                                                            sizeof(base_t));
+    }
+  }
+
+  void set_sycl_graph_ptr(
+      sycl::ext::oneapi::experimental::command_graph<
+          sycl::ext::oneapi::experimental::graph_state::modifiable>*
+          arg_graph) {
+    m_graph_ptr = arg_graph;
+  }
+
+  void set_sycl_graph_node_ptr(
+      std::optional<sycl::ext::oneapi::experimental::node>* arg_node) {
```

Comment:
A pointer to an optional?!

Reply:
That was the easiest way to adapt the Cuda/HIP implementation where the node object can be default constructed and checked for validity. This is not the case for `SYCL` which is why I wrap it in an `std::optional`.

Here, I basically need to set the pointer that initially points to an empty node which is then set through this class' interface. I can't store a reference since the value is only set in a member function but I could pass a reference instead (which is more or less the same).